### PR TITLE
 Flush uncommitted translog older than a configurable threshold 

### LINKED ISSUE #57610: Flush shards when uncommitted translog is old
Today we periodically flush a shard when the uncommitted translog size reaches 512MB (the default [flush threshold](https://github.com/elastic/elasticsearch/blob/ee7868d68733f195dc46926a7eab3d9dd7033ef4/server/src/main/java/org/elasticsearch/index/IndexSettings.java#L188)), or when it becomes [idle](https://github.com/elastic/elasticsearch/blob/725dda37ea570741595df25de2f86dc794cfec68/server/src/main/java/org/elasticsearch/indices/IndexingMemoryController.java#L403) (i.e., without indexing for 5 minutes). If the indexing rate is low but not low enough for a shard goes idle, then we will never flush it. @DaveCTurner suggested that we should also periodically flush a shard when the age of the uncommitted translog reaches a certain threshold to reduce the indexing memory and prevents recoveries from taking too long.

### PR DESCRIPTION
**Pull request by fcofdez**
Adds a periodic check to flush shards with translog files older than
`indices.memory.shard_uncommitted_translog_max_age` threshold.

Closes https://github.com/elastic/elasticsearch/issues/57610

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/indices/IndexingMemoryController.java [hunk 1]
```diff
@@ -83,6 +83,16 @@
         Property.NodeScope
     );
 
+    /**
+     * If the uncommitted translog files for a given shard reach this age, the shard will be flushed.
+     */
+    public static final Setting<TimeValue> SHARD_UNCOMMITTED_TRANSLOG_MAX_AGE_SETTING = Setting.timeSetting(
+        "indices.memory.shard_uncommitted_translog_max_age",
+        TimeValue.timeValueHours(1),
```
Comment (fcofdez) [inline server/src/main/java/org/elasticsearch/indices/IndexingMemoryController.java:91]: I'm not sure about the default and min values for this new configuration parameter, any feedback is welcomed.

File: server/src/main/java/org/elasticsearch/index/translog/Translog.java [hunk 2]
```diff
@@ -472,6 +472,26 @@ public long sizeInBytesByMinGen(long minGeneration) {
         }
     }
 
+    /**
+     * Returns the age of the oldest translog file with at least the given generation
+     */
+    public long getOldestTranslogAgeInMillisByMinGen(long minGeneration) {
```
Comment (dnhatn) [inline server/src/main/java/org/elasticsearch/index/translog/Translog.java:478]: Perhaps delegate `earliestLastModifiedAge` to `getOldestTranslogAgeInMillisByMinGen(-1L)`?

File: server/src/internalClusterTest/java/org/elasticsearch/indices/flush/FlushIT.java [hunk 3]
```diff
@@ -133,4 +137,38 @@ public void testFlushOnInactive() throws Exception {
             }
         }, 30, TimeUnit.SECONDS);
     }
+
+    public void testFlushOnceUncommittedTranslogReachesMaxAge() throws Exception {
+        final String indexName = "flush_on_old_translog";
+        List<String> dataNodes = internalCluster().startDataOnlyNodes(2, Settings.builder()
+            .put(IndexingMemoryController.SHARD_UNCOMMITTED_TRANSLOG_MAX_AGE_SETTING.getKey(), TimeValue.timeValueSeconds(3))
+            .put(IndexingMemoryController.SHARD_INACTIVE_TIME_SETTING.getKey(), TimeValue.timeValueDays(1)).build());
```
Comment (dnhatn) [inline server/src/internalClusterTest/java/org/elasticsearch/indices/flush/FlushIT.java:145]: Perhaps reduce `indices.memory.interval` to make this test run faster.

File: server/src/internalClusterTest/java/org/elasticsearch/indices/flush/FlushIT.java [hunk 4]
```diff
@@ -133,4 +137,38 @@ public void testFlushOnInactive() throws Exception {
...
+            .build()));
+        ensureGreen(indexName);
+        int numDocs = randomIntBetween(1, 50);
+        for (int i = 0; i < numDocs; i++) {
+            client().prepareIndex(indexName)
+                .setSource("f", "v".repeat((int) translogGenerationThreshold.getBytes())).get(); // force multiple translog generations
+        }
+
+        if (randomBoolean()) {
+            internalCluster().restartNode(randomFrom(dataNodes), new InternalTestCluster.RestartCallback());
```
Comment (dnhatn) [inline server/src/internalClusterTest/java/org/elasticsearch/indices/flush/FlushIT.java:164]: Do we need this condition?

File: server/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java [hunk 5]
```diff
@@ -446,7 +446,6 @@ protected final boolean isClosed() {
         return closed.get();
     }
 
-
```
Comment (dnhatn) [inline server/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java:449]: Unrelated change?

**Review by dnhatn — DISMISSED** [review]
I left some small comments, but LGTM.

**Review by ywelsch — COMMENTED** [review]
I wonder if we need some safeguarding, similar to what was done in `shouldPeriodicallyFlush`, as the information we rely on here could be off (we rely on the filesystem to provide correct timestamp info, and in case it might always return wrong stale values, we go flushing every 5 seconds). Also, I think we should avoid asking timestamp information for every open translog file every 5 seconds. Perhaps that info can be determined for TranslogReader when opening the file, and updated for TranslogWriter when we actively write. Finally, I wonder if we can fold this functionality together with `shard.inactive`. Maybe it's not worth distinguishing the case of "no active writes" and "trickling writes", and having separate configuration options just introduces unnecessary complexity.

**Comment by fcofdez** [comment]
> I wonder if we need some safeguarding, similar to what was done in shouldPeriodicallyFlush, as the information we rely on here could be off

That's a good point, I'll add a new safeguard for that case and some testing around that particular case 👍

> Perhaps that info can be determined for TranslogReader when opening the file, and updated for TranslogWriter when we actively write

👍 

>  Maybe it's not worth distinguishing the case of "no active writes" and "trickling writes", and having separate configuration options just introduces unnecessary complexity.

Do you mean changing the current parameter to a rate? meaning that we would consider a shard as inactive once is below some ops/time threshold?

**Comment by ywelsch** [comment]
> Do you mean changing the current parameter to a rate? meaning that we would consider a shard as inactive once is below some ops/time threshold?

No, I would rather like to unify both `shard.inactive` and `indices.memory.shard_uncommitted_translog_max_age` into a single configuration option. In case where a shard is inactive for 5 minutes, all items in the translog will have an age older than 5 minutes. That means that if `indices.memory.shard_uncommitted_translog_max_age` would be configured to 5 minutes, it would cover what `shard.inactive` is already doing today (+ also deal with low activity).

**Comment by fcofdez** [comment]
That would mean that we would flush every 5 minutes in other scenarios, i.e. the translog size in that time interval is < `index.translog.flush_threshold_size`. Is that something we want?
